pkg/sql/schemachanger/scbuild/table.go

```diff
@@ -293,18 +294,31 @@ func (b *buildContext) alterTableDropColumn(
 		}
 		panic(err)
 	}
-	// Check whether the column is being dropped.
+	var columnID descpb.ColumnID
+	columnIDFound := false
 	for _, n := range b.output {
 		switch col := n.Element().(type) {
-		case *scpb.Column:
-			if col.TableID != table.GetID() ||
-				n.Target.Direction != scpb.Target_DROP ||
-				col.Column.ColName() != t.Column {
-				continue
+		case *scpb.ColumnName:
+			if col.Name == t.Column.String() {
+				columnID = col.ColumnID
+				columnIDFound = true
+			}
+		}
+	}
+	if columnIDFound {
+		// Check whether the column is being dropped.
+		for _, n := range b.output {
```

Comment:
It seems like it's high time to add more functionality to the `buildContext` to make it more convenient to filter nodes in the `output`. The builder logic is overly verbose as it is.

I see you've introduced `forEachNodeOfType` and that's a great first step, I'm arguing that we should go further. In this case, perhaps something like a `forEachColumn(func(scpb.Column) error)` method that supports `iterutil.Done` would help, perhaps supplemented by a `FindColumn` method also. Errors are not returned by `forEachColumn` because it takes care of panicking all non-`iterutil.Done` errors.

Dealing with `reflect`, type casts, and so forth is distracting and awkward. I can see us needing these for each element type. These should probably be generated automatically.